c++/src/kj/compat/http.c++

```diff
@@ -3489,6 +3489,318 @@ WebSocketPipe newWebSocketPipe() {
 
 // =======================================================================================
 
+namespace {
+  inline kj::Maybe<StringPtr> advanceToNext(const StringPtr input, const char& c,
+      kj::Maybe<size_t>& advancedBy) {
+    // Given a string and character, return a StringPtr to the first non-whitespace character
+    // following the first instance of `char c`. The index of `char c` relative to `input` is stored
+    // in `advancedBy`.
+    kj::Maybe<StringPtr> result;
+    advancedBy = input.findFirst(c);
+    KJ_IF_MAYBE(index, advancedBy) {
+      // We found a delimiter, move to 1 past it, then skip over the whitespace.
+      auto startOfNext = const_cast<char*>((input.begin() + *index + 1));
+      result = skipSpace(startOfNext);
+    }
+    return result;
+  }
+
+  kj::Vector<kj::String> splitPartsWithCondition(const kj::StringPtr input, const char& delim,
+      kj::Maybe<StringPtr> condition = nullptr) {
+    // Given a string `input` and a delimiter `delim`, split the string into a vector of substrings,
+    // separated by the delimiter. Note that leading whitespace is stripped from each element.
+    //
+    // Optionally, the caller may pass a `condition` string, requesting that each substring starts
+    // with the value inside `condition`. This allows for filtering in 1 pass, rather than 2.
+    kj::Vector<kj::String> parts;
+
+    StringPtr cursor = input;
+    kj::Maybe<size_t> advanceCursor;
+    // How many characters (if any) were between cursor and the first instance of `delim`?
+
+    auto remainingParts = advanceToNext(cursor, delim, advanceCursor);
+    while (remainingParts != nullptr) {
+      auto part = kj::str(cursor.slice(0, KJ_REQUIRE_NONNULL(advanceCursor)));
+      KJ_IF_MAYBE(prefix, condition) {
+        // If the caller passed a condition, we filter before adding to the vector.
+        if (part.startsWith(*prefix)) {
+          parts.add(kj::mv(part));
+        }
+      } else {
+        parts.add(kj::mv(part));
+      }
+      cursor = KJ_REQUIRE_NONNULL(remainingParts).begin();
+      remainingParts = advanceToNext(cursor, delim, advanceCursor);
+    }
+    if (cursor.size() != 0) {
+      // If we have no more instances of `delim`, let's just add whatever remains.
+      KJ_IF_MAYBE(prefix, condition) {
+        if (cursor.startsWith(*prefix)) {
+          parts.add(kj::str(cursor));
+        }
+      } else {
+        parts.add(kj::str(cursor));
+      }
+    }
+
+    return parts;
+  }
+
+  inline kj::Maybe<kj::Exception> validateTokenTail(const StringPtr s) {
```

Comment:
Going to give some context on this as well. Suppose I wanted to verify the parameter `client_max_window_bits`. Previously, I was only checking if `parameter.startsWith("client_max_window_bits")`, which obviously doesn't account for anything _after_the token.

Currently, the `validateCompressionParameters()` function relies on `startsWith`, but it also uses `validateTokenTail()` to be certain there is no invalid stuff _after_ the token. It really boils down to this:
- `permessage-deflate` and `x_no_context_takeover` can be followed by `*LWS`.
- `x_max_window_bits` can be followed by:
    - `*LWS`
    - `*LWS` then an `"="`

We're just verifying that the stuff _after_ the token follows these rules.